src/shared/image-policy.c

```diff
@@ -0,0 +1,647 @@
+                if (a->policies[i].designator != b->policies[i].designator)
+                        return false;
+                if (a->policies[i].flags != b->policies[i].flags)
+                        return false;
+        }
+
+        return true;
+}
+
+int image_policy_equivalent(const ImagePolicy *a, const ImagePolicy *b) {
+
+        if (image_policy_default(a) != image_policy_default(b))
+                return false;
+
+        for (PartitionDesignator d = 0; d < _PARTITION_DESIGNATOR_MAX; d++) {
+                PartitionPolicyFlags f, w;
+
+                f = image_policy_get_exhaustively(a, d);
+                if (f < 0)
+                        return f;
+
+                w = image_policy_get_exhaustively(b, d);
+                if (w < 0)
+                        return w;
+
+                if (f != w)
+                        return false;
+        }
+
+        return true;
+}
+
+const ImagePolicy image_policy_allow = {
+        /* Allow policy */
+        .n_policies = 0,
+        .default_flags = PARTITION_POLICY_OPEN,
+};
+
+const ImagePolicy image_policy_deny = {
+        /* Allow policy */
+        .n_policies = 0,
+        .default_flags = PARTITION_POLICY_ABSENT,
+};
+
+const ImagePolicy image_policy_ignore = {
+        /* Allow policy */
+        .n_policies = 0,
+        .default_flags = PARTITION_POLICY_IGNORE,
+};
+
+const ImagePolicy image_policy_sysext = {
+        /* For system extensions, honour root file system, and /usr/ and ignore everything else. After all,
+         * we are only interested in /usr/ + /opt/ trees anyway, and that's really the only place they can
+         * be. */
+        .n_policies = 2,
+        .policies = {
+                { PARTITION_ROOT,     PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED|PARTITION_POLICY_ENCRYPTED|PARTITION_POLICY_UNPROTECTED|PARTITION_POLICY_ABSENT },
+                { PARTITION_USR,      PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED|PARTITION_POLICY_ENCRYPTED|PARTITION_POLICY_UNPROTECTED|PARTITION_POLICY_ABSENT },
+        },
+        .default_flags = PARTITION_POLICY_IGNORE,
+};
+
+const ImagePolicy image_policy_container = {
+        /* For systemd-nspawn containers we use all partitions, with the exception of swap */
+        .n_policies = 8,
+        .policies = {
+                { PARTITION_ROOT,     PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED|PARTITION_POLICY_ENCRYPTED|PARTITION_POLICY_UNPROTECTED|PARTITION_POLICY_ABSENT },
+                { PARTITION_USR,      PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED|PARTITION_POLICY_ENCRYPTED|PARTITION_POLICY_UNPROTECTED|PARTITION_POLICY_ABSENT },
+                { PARTITION_HOME,     PARTITION_POLICY_ENCRYPTED|PARTITION_POLICY_UNPROTECTED|PARTITION_POLICY_ABSENT },
+                { PARTITION_SRV,      PARTITION_POLICY_ENCRYPTED|PARTITION_POLICY_UNPROTECTED|PARTITION_POLICY_ABSENT },
+                { PARTITION_ESP,      PARTITION_POLICY_UNPROTECTED|PARTITION_POLICY_ABSENT },
```

Comment:
why are the esp/boot partition used for containers?

Reply:
When working with DDIs it's simply extremely useful if we can run "bootctl" inside a container to install it, or list boot menu items. Even if we don't use the pratitions for booting it's still extremely useful being able to manipulate/introspect the boot loader with nspawn.

I mean, I am sure we could have done this differently, i.e. run a stricter policy by default, and require people to open it up manually if they want to interact with boot loader stuff, but I think that ship has sailed, this always worked and I think compatibility in that sense matters too.

Hence I think we should run the tightest policies by default that are *definitely* not going to conflict with people's usecases. For example forbidding swap definitely is within the bounds of that.

Or making ESP/XBOOTLDR unavailable in sysext also makes sense, simply because sysext only covers /opt/ and /usr/, and esp/xbootldr is placed in /boot/ and /efi/, hence it's not possibly visible even if we mounted it...